src/qibo/quantum_info/metrics.py

```diff
@@ -0,0 +1,216 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+from scipy.linalg import sqrtm
+
+from qibo.config import PRECISION_TOL
+
+
+def purity(state):
+    """Purity of a quantum state :math:`\\rho`, which is given by :math:`\\text{Tr}(\\rho^{2})`.
+
+    Args:
+        state: state vector or density matrix.
+
+    Returns:
+        Purity of quantum state :math:`\\rho`.
+
+    """
+
+    if (
+        (len(state.shape) >= 3)
+        or (len(state) == 0)
+        or (len(state.shape) == 2 and state.shape[0] != state.shape[1])
+    ):
+        raise TypeError(
+            f"Object must have dims either (k,) or (k,k), but have dims {state.shape}."
+        )
```

Comment:
In qibo we have a custom way of raising errors which is handled by the logger, so that the version and datetime is logged with the error. This can be used as
```py
from qibo.config import raise_error

raise_error(TypeError, "error text")
```
If you decide to use this all errors need to be replaced.

Reply:
Changed it. Thanks.